[DOCU-2925] HTTPie cleanout (#5652)
Kong/docs.konghq.com: app/_src/gateway/production/access-control/enable-rbac.md

to achieve a flexible organization of teams and users in complex
hierarchies. 

## Use case

For the sake of example, let's say a given company has a {{site.base_gateway}}
cluster to be shared with two teams: teamA and teamB. While the Kong
clusters are shared among these teams, they want to be able to segment
their entities in such a way that management of entities in one team doesn't
disrupt operation in some other team. As shown in the
[Workspaces Page][workspaces-examples], such a use case is possible
with workspaces. On top of workspaces, though, each team wants to enforce
access control over their Workspace, which is possible with RBAC. 

To sum up, workspaces and RBAC are complementary: workspaces provide segmentation of
Admin API entities, while RBAC provides access control.

{:.note}

> > **Note:** The example responses in this guide are often partial excerpts 

Author: ```suggestion
> **Note:** The example responses in this guide are often excerpts of full responses,
```

focusing on the most relevant part of the response, as the full response can be very long.

## Bootstrapping the first RBAC user

The first RBAC user is called the super admin.

Kong recommends that you create a super admin
user before actually enforcing RBAC and restarting {{site.base_gateway}} 
with RBAC enabled.

{:.note}
> It's possible to create the first super admin at installation time.
If you chose this option, skip to [Enforcing RBAC](#enforcing-rbac).

> {{site.base_gateway}} ships with a set of default RBAC roles: `super-admin`,
> the `admin`, and `read-only`, which the task of creating a Super Admin user easy:

Author: ```suggestion
{{site.base_gateway}} ships with a set of default RBAC roles: `super-admin`,
the `admin`, and `read-only`, which makes the task of creating a super admin user easy:
```

    curl -i -X POST http://localhost:8001/rbac/users \
      -H Kong-Admin-Token:secureadmintoken \
      --data name=super-admin
    ```

    Response:
    ```json
    {
      "user_token": "M8J5A88xKXa7FNKsMbgLMjkm6zI2anOY",
      "id": "da80838d-49f8-40f6-b673-6fff3e2c305b",
      "enabled": true,
      "created_at": 1531009435000,
      "name": "super-admin"
    }
    ```

    As the `super-admin` username coincides with an existing `super-admin`
    role, it gets automatically added to the `super-admin` role. 

1. Confirm using the following command:

    ```sh
    curl -i -X GET http://localhost:8001/rbac/users/super-admin/roles
    ```

    Response:
    ```json
    {
      "roles": [
        {
          "comment": "Full access to all endpoints, across all workspaces",
          "created_at": 1531009724000,
          "name": "super-admin",
          "id": "b924ac91-e83f-4136-a5a4-4a7ff92594a8"
        }
      ],
      "user": {
        "created_at": 1531009858000,
        "id": "e6897cc0-0c34-4a9c-9f0b-cc65b4f04d68",
        "name": "super-admin",
        "enabled": true,
        "user_token": "vajeOlkybsn0q0VD9qw9B3nHYOErgY7b8"
      }
    }
    ```

## Enforcing RBAC

> As the `super-admin` user has just been created, the Kong admin may now
> restart Kong with RBAC enforced:

Author: ```suggestion
With the `super-admin` user created, the Kong admin can now restart 
{{site.base_gateway}} with RBAC enforced:
```

```
KONG_ENFORCE_RBAC=on kong restart
```

This is one of the possible ways of enforcing RBAC and restarting

> Kong. Another possibility is editing the Kong configuration file and

Author: ```suggestion
Kong. Another options is editing the {{site.base_gateway}} configuration file and
```

    ```

    Response:
    ```json
    {
      "name": "teamB",
      "created_at": 1531014143000,
      "id": "7dee8c56-c6db-4125-b87a-b508baa33c66"
    }
    ```

## Super admin creates one admin for each team

1. Create an admin for `teamA`:

    ```sh
    curl -i -X POST http://localhost:8001/teamA/rbac/users \
      --data name=adminA \
      --data user_token=exampletokenA \
      -H 'Kong-Admin-Token:vajeOlkbsn0q0VD9qw9B3nHYOErgY7b8'
    ```

    Response:
    ```json
    {
      "user_token": "qv1VLIpl8kHj7lC1QOKwRdCMXanqEDii",
      "id": "4d315ff9-8c1a-4844-9ea2-21b16204a154",
      "enabled": true,
      "created_at": 1531015165000,
      "name": "adminA"
    }
    ```

1. Create an admin for `teamB`:

    ```sh
    curl -i -X POST http://localhost:8001/teamB/rbac/users \
      --data name=adminB \
      --data user_token=exampletokenB \
      -H 'Kong-Admin-Token:vajeOlkbsn0q0VD9qw9B3nHYOErgY7b8'
    ```

    Response:
    ```json
    {
      "user_token": "IX5vHVgYqM40tLcctdmzRtHyfxB4ToYv",
      "id": "49641fc0-8c9d-4507-bc7a-2acac8f2903a",
      "enabled": true,
      "created_at": 1531015221000,
      "name": "adminB"
    }
    ```

> 1. With this, both of the teams have one admin and each admin can only be seen

Author: ```suggestion
1. Both of the teams now have one admin and each admin can only be seen
```

    ```sh
    curl -i -X GET http://localhost:8001/teamA/rbac/users \
      -H 'Kong-Admin-Token:vajeOlkbsn0q0VD9qw9B3nHYOErgY7b8'
    ```

    Response:
    ```json
    {
      "total": 1,
      "data": [
        {
          "created_at": 1531014784000,
          "id": "1faaacd1-709f-4762-8c3e-79f268ec8faf",
          "name": "adminA",
          "enabled": true,
          "user_token": "n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG"
        }
      ]
    }
    ```

    Similarly, workspace `teamB` only shows its own admin:

    ```sh
    curl -i -X GET http://localhost:8001/teamB/rbac/users \
      -H 'Kong-Admin-Token:vajeOlkbsn0q0VD9qw9B3nHYOErgY7b8'
    ```

    Response:
    ```json
    {
      "total": 1,
      "data": [
        {
          "created_at": 1531014805000,
          "id": "3a829408-c1ee-4764-8222-2d280a5de441",
          "name": "adminB",
          "enabled": true,
          "user_token": "C8b6kTTN10JFyU63ORjmCQwVbvK4maeq"
        }
      ]
    }
    ```

## Super admin creates admin roles for teams

The super admin is now done creating RBAC admin users for each team. The next
task is to create admin roles that will effectively grant permissions to admin
users.

> The admin role must have access to all of the Admin API, restricted to their

Author: ```suggestion
The admin role must have access to all of the Admin API, restricted to its
```

          "created_at": 1531017322000,
          "role_id": "d40e61ab-8dad-4ef2-a48b-d11379f7b8d1",
          "actions": [
            "delete",
            "create",
            "update",
            "read"
          ],
          "negative": false,
          "workspace": "teamA"
        }
      ]
    }
    ```

1. Add the `adminA` user to the admin role in their workspace:

    ```sh
    curl -i -X POST http://localhost:8001/teamA/rbac/users/adminA/roles/ \
      --data roles=admin \
      -H 'Kong-Admin-Token:vajeOlkbsn0q0VD9qw9B3nHYOErgY7b8'
    ```

    Response: 

    ```json
    {
      "roles": [
          {
              "created_at": 1685551877,
              "id": "42809ada-650c-4575-b0a0-d464a64ffb70",
              "name": "admin",
              "ws_id": "9dc7adbb-9b64-4121-bf76-653cf5871bc2"
          }
      ],
      "user": {
          "comment": "null",
          "created_at": 1685552809,
          "enabled": true,
          "id": "bca4e390-fbbf-4a46-b55d-f4642efc14bb",
          "name": "adminA",
          "user_token": "$2b$09$oLyKTIDuKriPZ.SD5wYtxeMclGYNDn4udJkQG0NGx/Aq3j9j/tWsa",
          "user_token_ident": "0ebb5"
            }
    }
    ```

> 1. With these steps, Team A's admin user is now able to manage their team. To

Author: ```suggestion
1. Team A's admin user is now able to manage their team. To
```

validate that, let's try to list RBAC users in Team B using Team A's admin
user token:

    ```sh
    curl -i -X GET http://localhost:8001/teamB/rbac/users \
      -H 'Kong-Admin-Token:n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG'
    ```

    Notice that you can't access the endpoint:
    ```json
    {
      "message": "Invalid RBAC credentials"
    }
    ```

> 1. Now try to list RBAC users in Team A's workspace:

Author: ```suggestion
1. Now try listing RBAC users in Team A's workspace:
```

    ```sh
    curl -i -X GET http://localhost:8001/teamA/rbac/users \
      -H 'Kong-Admin-Token:n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG'
    ```

    Response:
    ```json
    {
      "total": 1,
      "data": [
        {
          "created_at": 1531014784000,
          "id": "1faaacd1-709f-4762-8c3e-79f268ec8faf",
          "name": "adminA",
          "enabled": true,
          "user_token": "n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG"
        }
      ]
    }
    ```

If the same procedure is repeated for Team B, they will end up with
a similar set up, with an admin role and an admin user, both restricted to the
team's workspace.

The super admin's job is now done. 

> Individual team admins are now able to set up their teams users and entities!

Author: ```suggestion
Individual team admins are now able to set up their team's users and entities!
```

## Team admins create regular users

From this point on, team admins are able to drive the process. The next
step is to create team users, such as engineers that are part of Team A or B. 
Let's go ahead and do that, using Admin A's user token.

Before regular users can be created, a role needs to be available for them.

> This role needs to have permissions to all of Admin API endpoints, except
> RBAC and workspaces. Regular users don't need access to these endpoints in general
> and if they do, the admin can grant them.

Author: ```suggestion
This role needs to have permissions to all of the Admin API endpoints, except
RBAC and workspaces. Regular users don't need access to these endpoints,
and if they do, the admin can grant them individually.
```

    ```sh
    curl -i -X POST http://localhost:8001/teamA/rbac/roles/users/endpoints/ \
      --data 'endpoint=/rbac/*' \
      --data 'workspace=teamA' \
      --data 'actions=*' \
      -H 'Kong-Admin-Token:n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG'
    ```

    Response:
    ```json
    {
      "endpoint": "/rbac/*",
      "created_at": 1531020744000,
      "role_id": "9846b92c-6820-4741-ac31-425b3d6abc5b",
      "actions": [
        "delete",
        "create",
        "update",
        "read"
      ],
      "negative": true,
      "workspace": "teamA"
    }
    ```

    Filter out workspaces endpoints:

    ```sh
    curl -i -X POST http://localhost:8001/teamA/rbac/roles/users/endpoints/ \
      --data 'endpoint=/workspaces/*' \
      --data 'workspace=teamA' \
      --data 'actions=*' \
      -H 'Kong-Admin-Token:n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG'
    ```

    Response:
    ```json
    {
      "endpoint": "/workspaces/*",
      "created_at": 1531020778000,
      "role_id": "9846b92c-6820-4741-ac31-425b3d6abc5b",
      "actions": [
        "delete",
        "create",
        "update",
        "read"
      ],
      "negative": true,
      "workspace": "teamA"
    }
    ```

{:.important}

> > **Important**: As explained in the [Wildcards in Permissions](#wildcards-in-permissions)
> section, the meaning of `*` is not the expected generic globbing one might
> be used to. As such, `/rbac/*` or `/workspaces/*` do not match all of the

Author: ```suggestion
> **Important**: As explained in the [Wildcards in Permissions](#wildcards-in-permissions)
section, the meaning of `*` is not the same as generic globbing.
 As such, `/rbac/*` or `/workspaces/*` do not match all of the
```

> RBAC and Workspaces endpoints. For example, to cover all of the RBAC API,
> one would have to define permissions for the following endpoints:

Author: ```suggestion
RBAC and Workspaces endpoints. For example, to cover all of the RBAC API,
you would have to define permissions for the following endpoints:
```

    }
    ```

1. List currently enabled plugins:

    ```sh
    curl -i -X GET http://localhost:8001/teamA/plugins \
      -H 'Kong-Admin-Token:dNeYvYAwvjOJdoReVJZXF8vLBXQioKkI'
    ```

    Response:
    ```json
    {
      "total": 1,
      "data": [
        {
          "created_at": 1531021732000,
          "config": {
            "key_in_body": false,
            "run_on_preflight": true,
            "anonymous": "",
            "hide_credentials": false,
            "key_names": [
              "apikey"
            ]
          },
          "id": "cdc85ef0-804b-4f92-aafd-3ff58512e445",
          "name": "key-auth",
          "enabled": true
        }
      ]
    }
    ```

This ends the use case tutorial, which demonstrates the power of RBAC and
workspaces with a real-world scenario.

## Entity-level RBAC

In addition to endpoint permissions, RBAC in {{site.base_gateway}} supports entity-level
permissions, meaning that particular entities, identified by their unique ID,
can be allowed or disallowed access in a role.

RBAC is [enforced](#enforcing-rbac) with the `enforce_rbac`
configuration directive, or with its `KONG_ENFORCE_RBAC` environment variable

> counterpart. Such directive is an enum, with 4 possible values:

Author: ```suggestion
counterpart. The directive is an enum, with the following possible values:
```

- `on`: Applies endpoint-levelaccess control
- `entity`: Applies **only** Entity-level access control
- `both`: Applies **both Endpoint and Entity level access control**
- `off`: disables RBAC enforcement

> If set to `entity` or `both`, Kong will enforce entity-level

Author: ```suggestion
If set to `entity` or `both`, Kong enforces entity-level
```

        "created_at": 1531065373000,
        "id": "4d87bf78-5824-4756-b0d0-ceaa9bd9b2d5",
        "name": "qux",
        "enabled": true,
        "user_token": "sUnv6uBehM91amYRNWESsgX3HzqoBnR5"
      }
    }
    ```

1. Check that the correct permissions are listed for `qux`:

    ```sh
    curl -i -X GET http://localhost:8001/teamA/rbac/users/qux/permissions \
      -H 'Kong-Admin-Token:n5bhjgv0speXp4N7rSUzUj8PGnl3F5eG'
    ```

    Response:
    ```json
    {
      "entities": {
        "d25afc46-dc59-48b2-b04f-d3ebe19f6d4b": {
          "actions": [
            "read"
          ],
          "negative": false
        },
        "3ed24101-19a7-4a0b-a10f-2f47bcd4ff43": {
          "actions": [
            "read"
          ],
          "negative": false
        }
      },
      "endpoints": {}
    }
    ```
    
    `qux` should have two entity permissions and no endpoint permissions.

Admin A is done setting up `qux`, and `qux `can now use their user token to read
the two entities over Kong's admin API.

Let's assume that Admin A [enabled entity-level enforcement](#enforcing-rbac) as well.
Note that as `qux` has **no endpoint-level permissions**. If both endpoint and

> entity-level enforcement is enabled, `qux` won't  be able to read their entities, as

Author: ```suggestion
entity-level enforcement is enabled, `qux` won't be able to read their entities, as
```
